lib/src/trailer.rs

```diff
@@ -0,0 +1,264 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Parsing trailers from commit messages.
+
+use itertools::Itertools as _;
+
+/// A key-value pair representing a trailer in a commit message, of the
+/// form `Key: Value`.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Trailer {
+    /// trailer key
+    pub key: String,
+    /// normalized trailer value
+    pub value: String,
+}
+
+/// Parse the trailers from a commit message; these are simple key-value
+/// pairs, separated by a colon, describing extra information in a commit
+/// message; an example is the following:
+///
+/// ```text
+/// chore: update itertools to version 0.14.0
+///
+/// Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
+/// tempor incididunt ut labore et dolore magna aliqua.
+///
+/// Co-authored-by: Alice <alice@example.com>
+/// Co-authored-by: Bob <bob@example.com>
+/// Reviewed-by: Charlie <charlie@example.com>
+/// Change-Id: I1234567890abcdef1234567890abcdef12345678
+/// ```
+///
+/// In this case, there are four trailers: two `Co-authored-by` lines, one
+/// `Reviewed-by` line, and one `Change-Id` line.
+pub fn parse_description_trailers(body: &str) -> Vec<Trailer> {
+    let (trailer, blank) = parse_trailers_and_blank(body);
+    if blank {
+        trailer
+    } else {
+        // no blank found, this means there was a single paragraph, so whatever
+        // was found can't come from the trailer
+        vec![]
+    }
+}
+
+/// Parse the trailers from a trailer paragraph. This function behaves like
+/// `parse_description_trailer`, except that it doesn't expect the body to
+/// contain several paragraphs.
+pub fn parse_trailers(body: &str) -> Vec<Trailer> {
+    let (trailer, _) = parse_trailers_and_blank(body);
+    trailer
+}
+
+fn parse_trailers_and_blank(body: &str) -> (Vec<Trailer>, bool) {
+    // a trailer always comes at the end of a message; we can split the message
+    // by newline, but we need to immediately reverse the order of the lines
+    // to ensure we parse the trailer in an unambiguous manner; this avoids cases
+    // where a colon in the body of the message is mistaken for a trailer
+    let lines = body.trim_end().lines().rev();
+    let trailer_re =
+        regex::Regex::new(r"^([a-zA-Z0-9-]+) *: +(.+) *$").expect("trailer regex should be valid");
+    let mut trailer: Vec<Trailer> = Vec::new();
+    let mut multiline_value: Vec<&str> = Vec::new();
+    let mut found_blank = false;
+    for line in lines {
+        if line.starts_with(' ') {
+            multiline_value.push(line.trim());
+        } else if let Some(caps) = trailer_re.captures(line) {
+            let key = caps[1].trim().to_string();
+            multiline_value.push(caps.get(2).unwrap().as_str());
+            let value = multiline_value.iter().rev().join(" ");
```

Comment:
nit: perhaps, multi-line text should be concatenated with `\n`? It's nice if the original trailer text can be reconstructed from a parsed `Trailer` object (ignoring stripped whitespace.)

Reply:
It shouldn't when looking at the parsed value, but it would indeed be nice to keep the user formatting.

I was thinking of keeping the text generated by the user in a `text` field in the `Trailer` struct. That way, `key` and `value` could be used to compare trailers that have been formatted differently, and the unmodified `text` would be used to add the trailer to the description.

I already began to do that, but then I wasn't sure how `key`, `value`, and `text` should behave when one of them is updated. That's something that could be decided later though.
If you think that preserving the formatting is important, I can work on that.